Optimize the instruction fetch path

yy214123 opened the pull request:
This improves the instruction-fetch path by adding new cache structures and enhancing the observability of fetch behavior.

**Key Improvements**

1. Replaces the previous 1-entry direct-mapped design with a 2-entry hash-indexed TLB.
2. Adds a direct-mapped instruction cache that stores recently fetched instructions and provides a fast hit path before triggering address translation.
3. Provides finer-grained fetch statistics to support and validate the above architectural changes.

Here’s the updated performance comparison:

| | 1-entry TLB | 2-entry TLB | + 64KB Direct-Mapped I-cache
-- | -- | -- | --
Fetch hit | 95,971,425 | 96,757,879 | 97,706,775
Fetch miss | 4,028,576 | 3,242,122 (-19.5%) | 1,026,354 (-68%)
Fetch hit rate | 95.97% | 96.76% | 98.96%
Load hit rate | 93.15% | 93.15% | 93.15%
Store hit rate | 95.57% | 95.57% | 95.57%

jserv commented:
> Also includes clang-format fixes for several expressions.

Don't do this. Always indent with clang-format-18.

Avoid inserting screenshots that contain only plain text, as visually impaired individuals may be unable to perceive or understand their content.

jserv commented on riscv.h:
```diff
@@ -69,7 +69,26 @@ typedef struct {
+#define IC_INDEX_BITS (__builtin_ctz((IC_BLOCKS)))
+#define IC_INDEX_MASK (IC_BLOCKS - 1)
+#define IC_BLOCK_MASK (IC_BLOCKS_SIZE - 1)
+#define RV_PAGE_MASK (RV_PAGE_SIZE - 1)
```
Add proper comments.

yy214123 replied:
Done. Added comments as suggested, thanks for the review!

yy214123 commented on riscv.h:
```diff
@@ -69,7 +69,26 @@ typedef struct {
+
+typedef struct {
+    ic_block_t block[IC_BLOCKS];
+} ic_t;
```
Done, rename to `icache`.

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on riscv.h:
```diff
@@ -75,7 +75,54 @@ typedef struct {
+#define IC_BLOCKS_SIZE 256
+#define IC_BLOCKS 256
+#define IC_OFFSET_BITS (__builtin_ctz((IC_BLOCKS_SIZE)))
+#define IC_INDEX_BITS (__builtin_ctz((IC_BLOCKS)))
```
Use `ICACHE_` as prefix.

jserv reviewed (changes requested):
Unify the naming scheme.

jserv reviewed (changes requested):
Rebase the latest 'master' branch and resolve build errors.

jserv reviewed (changes requested):
Squash commits and refine commit message.

visitorckw reviewed (changes requested):
This series appears to contain several "fix-up," "refactor," or "build-fix" commits that correct or adjust a preceding patch.

To maintain a clean history and ensure the project is bisectable, each patch in a series should be complete and correct on its own.

yy214123 commented:
> ...
> 
> To maintain a clean history and ensure the project is bisectable, each patch in a series should be complete and correct on its own.

I've finished cleaning up the commits and split them into three functionally independent patches.


> Please ensure that when you quote-reply to others' comments, you do not translate the quoted text into any language other than English.

Thank you for pointing that out. I've corrected the mistake and will be more careful in the future.

jserv commented on riscv.h:
```diff
@@ -75,7 +75,55 @@ typedef struct {
 typedef struct __vm_internel vm_t;
 
+/* IC_BLOCKS_SIZE: Size of one instruction-cache block (line).
+ * IC_BLOCKS: Number of blocks (lines) in the instruction cache.
```
Refine macro names.

yy214123 replied:
I've fixed the missing comments and rebased onto the latest master branch.

yy214123 commented on main.c:
```diff
@@ -1024,8 +1024,12 @@ static void print_mmu_cache_stats(vm_t *vm)
+        for (int i = 0; i < 2; i++) {
+            fetch_hits += hart->cache_fetch[i].hits;
+            fetch_misses += hart->cache_fetch[i].misses;
+        }
```
Could I use this two-line expression?
```diff
+   fetch_hits = hart->cache_fetch[1].hits + hart->cache_fetch[2].hits
+   fetch_misses = hart->cache_fetch[1].misses + hart->cache_fetch[2].misses
```

visitorckw commented on riscv.h:
```diff
@@ -75,7 +75,55 @@ typedef struct {
+ *   - index bits        = log2(ICACHE_BLOCKS)
+ *
+ * For power-of-two values, log2(x) equals the number of trailing zero bits in
+ * x. Therefore, we use __builtin_ctz(x) (count trailing zeros) to compute these
```
Sorry, I missed that.
I don't have a strong opinion on this, so both options are fine with me.

yy214123 commented on riscv.h:
```diff
@@ -75,7 +75,51 @@ typedef struct {
+ * instruction cache (IC), ensuring full block compatibility.
+ * However, the number of blocks is smaller, allowing the VC to store
+ * a few recently evicted cache lines to reduce conflict misses.
+ */
```
I enhanced the description of the interaction between the victim cache and the instruction cache. 
How does this version look?

```c
/* VCACHE_BLOCKS_SIZE: Size of one victim-cache block (line).
 * VCACHE_BLOCKS: Number of blocks (lines) in the victim cache.
 *
 * The victim cache is implemented as a small, fully associative cache. 
 * It is designed to serve as a temporary buffer for instruction cache blocks 
 * that were recently evicted from the instruction cache. 
 *  
 * Upon an instruction cache miss, the system first checks the victim cache 
 * for the corresponding data. If the data is found (a victim cache hit), 
 * the instruction cache block and the victim cache block are swapped. 
 * Conversely, when the instruction cache is being filled with new data, 
 * the evicted old data from the instruction cache block is simultaneously 
 * placed into the victim cache.
 */
```

jserv commented:
> Found that the victim cache implementation still has significant room for improvement:

What are you going to proceed with?

jserv commented:
> Does this approach make sense to you, or would you suggest a different direction?

When in doubt, measure before you choose a method.

yy214123 commented on main.c:
```diff
@@ -1047,14 +1069,32 @@ static void print_mmu_cache_stats(vm_t *vm)
+        fprintf(stderr, "\n=== Introduction Cache Statistics ===\n");
+        fprintf(stderr, "  Total access:  %12llu\n", access_total);
+        fprintf(stderr, "  Icache hits:   %12llu (%.2f%%)\n", fetch_hits_icache,
+                (fetch_hits_icache * 100.0) / access_total);
```
Fixed in 8bfe59c

yy214123 commented on main.c:
```diff
@@ -1047,14 +1069,32 @@ static void print_mmu_cache_stats(vm_t *vm)
+        fprintf(stderr,
+                "   ├ Vcache hits:    %8llu (%.2f%% of Icache misses)\n",
+                fetch_hits_vcache,
+                (fetch_hits_vcache * 100.0) / fetch_misses_icache,
```
Fixed in 8bfe59c

jserv commented:
The measurements reveal that 98% of the  benefit comes from a simple 2-entry TLB cache, not from the elaborate 64KB I-cache + 16-block victim cache with LRU. The working set fits comfortably in 64KB. The victim cache is solving a problem that doesn't exist. These are compulsory misses, not conflict misses.

LRU Overhead Cost:
```
  // O(N) linear scan on EVERY I-cache miss (6.19M times):
  for (int i = 1; i < 16; i++) {  // 16 comparisons each
      if (vm->icache.v_used[i] < lru_min) { ... }
  }
```
- 6.19M misses × 16 comparisons = 99M operations
- Saves only 405K victim cache hits

jserv commented on riscv.c:
```diff
@@ -361,10 +371,62 @@ static void mmu_fence(hart_t *vm, uint32_t insn UNUSED)
+    vm->cache_fetch[index].icache_misses++;
+#endif
+
+    uint32_t vcache_key = addr >> ICACHE_OFFSET_BITS;
```
Fragile tag calculation:
* Works by accident, obscures intent
* Should explicitly store block_id instead of reconstructing

yy214123 commented on riscv.h:
```diff
@@ -36,8 +36,11 @@ typedef struct {
-    uint64_t misses;
+    uint64_t total_fetch;
+    uint64_t tlb_hits;
+    uint64_t tlb_misses;
```
This has been completed.

yy214123 commented:
> Update the descriptions of this pull request.

I have updated the title and description at the top.

yy214123 commented:
> Is this pull request ready to proceed?

Yes, I am currently keeping only the simple implementations that can effectively bring improvements.

jserv commented:
Thank @yy214123 for contributing!
